File: dotnet/src/Planners/Planners.OpenAI/Stepwise/FunctionCallingStepwisePlanner.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Diagnostics.CodeAnalysis;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using Json.More;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;
using Microsoft.SemanticKernel.ChatCompletion;
using Microsoft.SemanticKernel.Connectors.OpenAI;

namespace Microsoft.SemanticKernel.Planning;

/// <summary>
/// A planner that uses OpenAI function calling in a stepwise manner to fulfill a user goal or question.
/// </summary>
public sealed class FunctionCallingStepwisePlanner
{
    /// <summary>
    /// Initialize a new instance of the <see cref="FunctionCallingStepwisePlanner"/> class.
    /// </summary>
    /// <param name="options">The planner options.</param>
    public FunctionCallingStepwisePlanner(
        FunctionCallingStepwisePlannerOptions? options = null)
    {
        this._options = options ?? new();
        this._generatePlanYaml = this._options.GetInitialPlanPromptTemplate?.Invoke() ?? EmbeddedResource.Read("Stepwise.GeneratePlan.yaml");
        this._stepPrompt = this._options.GetStepPromptTemplate?.Invoke() ?? EmbeddedResource.Read("Stepwise.StepPrompt.txt");
        this._options.ExcludedPlugins.Add(StepwisePlannerPluginName);
    }

    /// <summary>
    /// Execute a plan
    /// </summary>
    /// <param name="kernel">The <see cref="Kernel"/> containing services, plugins, and other state for use throughout the operation.</param>
    /// <param name="question">The question to answer</param>
    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
    /// <returns>Result containing the model's response message and chat history.</returns>
    public Task<FunctionCallingStepwisePlannerResult> ExecuteAsync(
        Kernel kernel,
        string question,
        CancellationToken cancellationToken = default)
    {
        var logger = kernel.LoggerFactory.CreateLogger(this.GetType()) ?? NullLogger.Instance;

        return PlannerInstrumentation.InvokePlanAsync(
            static (FunctionCallingStepwisePlanner plan, Kernel kernel, string? question, CancellationToken cancellationToken)
                => plan.ExecuteCoreAsync(kernel, question!, cancellationToken),
            this, kernel, question, logger, cancellationToken);
    }

    #region private

    private async Task<FunctionCallingStepwisePlannerResult> ExecuteCoreAsync(
        Kernel kernel,
        string question,
        CancellationToken cancellationToken = default)
    {
        Verify.NotNullOrWhiteSpace(question);
        Verify.NotNull(kernel);
        IChatCompletionService chatCompletion = kernel.GetRequiredService<IChatCompletionService>();
        ILoggerFactory loggerFactory = kernel.LoggerFactory;
        ILogger logger = loggerFactory.CreateLogger(this.GetType()) ?? NullLogger.Instance;
        var promptTemplateFactory = new KernelPromptTemplateFactory(loggerFactory);
        var stepExecutionSettings = this._options.ExecutionSettings ?? new OpenAIPromptExecutionSettings();

        // Clone the kernel so that we can add planner-specific plugins without affecting the original kernel instance
        var clonedKernel = kernel.Clone();
        clonedKernel.ImportPluginFromType<UserInteraction>();

        // Create and invoke a kernel function to generate the initial plan
        var initialPlan = await this.GeneratePlanAsync(question, clonedKernel, logger, cancellationToken).ConfigureAwait(false);

        var chatHistoryForSteps = await this.BuildChatHistoryForStepAsync(question, initialPlan, clonedKernel, promptTemplateFactory, cancellationToken).ConfigureAwait(false);

        for (int i = 0; i < this._options.MaxIterations; i++)
        {
            // sleep for a bit to avoid rate limiting
            if (i > 0)
            {
                await Task.Delay(this._options.MinIterationTimeMs, cancellationToken).ConfigureAwait(false);
            }

            // For each step, request another completion to select a function for that step
            chatHistoryForSteps.AddUserMessage(StepwiseUserMessage);
            var chatResult = await this.GetCompletionWithFunctionsAsync(chatHistoryForSteps, clonedKernel, chatCompletion, stepExecutionSettings, logger, cancellationToken).ConfigureAwait(false);
            chatHistoryForSteps.Add(chatResult);

            // Check for function response
            if (!this.TryGetFunctionResponse(chatResult, out IReadOnlyList<OpenAIFunctionToolCall>? functionResponses, out string? functionResponseError))
            {
                // No function response found. Either AI returned a chat message, or something went wrong when parsing the function.
                // Log the error (if applicable), then let the planner continue.
                if (functionResponseError is not null)
                {
                    chatHistoryForSteps.AddUserMessage(functionResponseError);
                }
                continue;
            }

            // Check for final answer in the function response
            foreach (OpenAIFunctionToolCall functionResponse in functionResponses)
            {
                if (this.TryFindFinalAnswer(functionResponse, out string finalAnswer, out string? finalAnswerError))
                {
                    if (finalAnswerError is not null)
                    {
                        // We found a final answer, but failed to parse it properly.
                        // Log the error message in chat history and let the planner try again.
                        chatHistoryForSteps.AddUserMessage(finalAnswerError);
                        continue;
                    }

                    // Success! We found a final answer, so return the planner result.
                    return new FunctionCallingStepwisePlannerResult
                    {
                        FinalAnswer = finalAnswer,
                        ChatHistory = chatHistoryForSteps,
                        Iterations = i + 1,
                    };
                }
            }

            // Look up function in kernel
            foreach (OpenAIFunctionToolCall functionResponse in functionResponses)
            {
                if (clonedKernel.Plugins.TryGetFunctionAndArguments(functionResponse, out KernelFunction? pluginFunction, out KernelArguments? arguments))
                {
                    try
                    {
                        // Execute function and add to result to chat history
                        var result = (await clonedKernel.InvokeAsync(pluginFunction, arguments, cancellationToken).ConfigureAwait(false)).GetValue<object>();
                        chatHistoryForSteps.AddMessage(AuthorRole.Tool, ParseObjectAsString(result), metadata: new Dictionary<string, object?>(1) { { OpenAIChatMessageContent.ToolIdProperty, functionResponse.Id } });
                    }
                    catch (Exception ex) when (!ex.IsCriticalException())
                    {
                        chatHistoryForSteps.AddMessage(AuthorRole.Tool, ex.Message, metadata: new Dictionary<string, object?>(1) { { OpenAIChatMessageContent.ToolIdProperty, functionResponse.Id } });
                        chatHistoryForSteps.AddUserMessage($"Failed to execute function {functionResponse.FullyQualifiedName}. Try something else!");
                    }
                }
                else
                {
                    chatHistoryForSteps.AddUserMessage($"Function {functionResponse.FullyQualifiedName} does not exist in the kernel. Try something else!");
                }
            }
        }

        // We've completed the max iterations, but the model hasn't returned a final answer.
        return new FunctionCallingStepwisePlannerResult
        {
            FinalAnswer = string.Empty,
            ChatHistory = chatHistoryForSteps,
            Iterations = this._options.MaxIterations,
        };
    }

    private async Task<ChatMessageContent> GetCompletionWithFunctionsAsync(
        ChatHistory chatHistory,
        Kernel kernel,
        IChatCompletionService chatCompletion,
        OpenAIPromptExecutionSettings openAIExecutionSettings,
        ILogger logger,
        CancellationToken cancellationToken)
    {
        openAIExecutionSettings.ToolCallBehavior = ToolCallBehavior.EnableKernelFunctions;

        await this.ValidateTokenCountAsync(chatHistory, kernel, logger, openAIExecutionSettings, cancellationToken).ConfigureAwait(false);
        return await chatCompletion.GetChatMessageContentAsync(chatHistory, openAIExecutionSettings, kernel, cancellationToken).ConfigureAwait(false);
    }

    private async Task<string> GetFunctionsManualAsync(Kernel kernel, ILogger logger, CancellationToken cancellationToken)
    {
        return await kernel.Plugins.GetJsonSchemaFunctionsManualAsync(this._options, null, logger, false, OpenAIFunction.NameSeparator, cancellationToken).ConfigureAwait(false);
    }

    // Create and invoke a kernel function to generate the initial plan
    private async Task<string> GeneratePlanAsync(string question, Kernel kernel, ILogger logger, CancellationToken cancellationToken)
    {
        var generatePlanFunction = kernel.CreateFunctionFromPromptYaml(this._generatePlanYaml);
        string functionsManual = await this.GetFunctionsManualAsync(kernel, logger, cancellationToken).ConfigureAwait(false);
        var generatePlanArgs = new KernelArguments
        {
            [NameDelimiterKey] = OpenAIFunction.NameSeparator,
            [AvailableFunctionsKey] = functionsManual,
            [GoalKey] = question
        };
        var generatePlanResult = await kernel.InvokeAsync(generatePlanFunction, generatePlanArgs, cancellationToken).ConfigureAwait(false);
        return generatePlanResult.GetValue<string>() ?? throw new KernelException("Failed get a completion for the plan.");
    }

    private async Task<ChatHistory> BuildChatHistoryForStepAsync(
        string goal,
        string initialPlan,
        Kernel kernel,
        KernelPromptTemplateFactory promptTemplateFactory,
        CancellationToken cancellationToken)
    {
        var chatHistory = new ChatHistory();

        // Add system message with context about the initial goal/plan
        var arguments = new KernelArguments
        {
            [GoalKey] = goal,
            [InitialPlanKey] = initialPlan
        };
        var systemMessage = await promptTemplateFactory.Create(new PromptTemplateConfig(this._stepPrompt)).RenderAsync(kernel, arguments, cancellationToken).ConfigureAwait(false);

        chatHistory.AddSystemMessage(systemMessage);

        return chatHistory;
    }

    private bool TryGetFunctionResponse(ChatMessageContent chatMessage, [NotNullWhen(true)] out IReadOnlyList<OpenAIFunctionToolCall>? functionResponses, out string? errorMessage)
    {
        OpenAIChatMessageContent? openAiChatMessage = chatMessage as OpenAIChatMessageContent;
        Verify.NotNull(openAiChatMessage, nameof(openAiChatMessage));

        functionResponses = null;
        errorMessage = null;
        try
        {
            functionResponses = openAiChatMessage.GetOpenAIFunctionToolCalls();
        }
        catch (JsonException)
        {
            errorMessage = "That function call is invalid. Try something else!";
        }

        return functionResponses is { Count: > 0 };
    }

    private bool TryFindFinalAnswer(OpenAIFunctionToolCall functionResponse, out string finalAnswer, out string? errorMessage)
    {
        finalAnswer = string.Empty;
        errorMessage = null;

        if (functionResponse.PluginName == "UserInteraction" && functionResponse.FunctionName == "SendFinalAnswer")
        {
            if (functionResponse.Arguments is { Count: > 0 } arguments && arguments.TryGetValue("answer", out object? valueObj))
            {
                finalAnswer = ParseObjectAsString(valueObj);
            }
            else
            {
                errorMessage = "Returned answer in incorrect format. Try again!";
            }
            return true;
        }
        return false;
    }

    private static string ParseObjectAsString(object? valueObj)
    {
        string resultStr = string.Empty;

        if (valueObj is RestApiOperationResponse apiResponse)
        {
            resultStr = apiResponse.Content as string ?? string.Empty;
        }
        else if (valueObj is string valueStr)
        {
            resultStr = valueStr;
        }
        else if (valueObj is JsonElement valueElement)
        {
            if (valueElement.ValueKind == JsonValueKind.String)
            {
                resultStr = valueElement.GetString() ?? "";
            }
            else
            {
                resultStr = valueElement.ToJsonString();
            }
        }
        else
        {
            resultStr = JsonSerializer.Serialize(valueObj);
        }

        return resultStr;
    }

    private async Task ValidateTokenCountAsync(
        ChatHistory chatHistory,
        Kernel kernel,
        ILogger logger,
        OpenAIPromptExecutionSettings openAIExecutionSettings,
        CancellationToken cancellationToken)
    {
        if (this._options.MaxPromptTokens is not null)
        {
            string functionManual = string.Empty;

            // If using functions, get the functions manual to include in token count estimate
            if (openAIExecutionSettings.ToolCallBehavior == ToolCallBehavior.EnableKernelFunctions)
            {
                functionManual = await this.GetFunctionsManualAsync(kernel, logger, cancellationToken).ConfigureAwait(false);
            }

            var tokenCount = chatHistory.GetTokenCount(additionalMessage: functionManual);
            if (tokenCount >= this._options.MaxPromptTokens)
            {
                throw new KernelException("ChatHistory is too long to get a completion. Try reducing the available functions.");
            }
        }
    }

    /// <summary>
    /// The options for the planner
    /// </summary>
    private readonly FunctionCallingStepwisePlannerOptions _options;

    /// <summary>
    /// The prompt YAML for generating the initial stepwise plan.
    /// </summary>
    private readonly string _generatePlanYaml;

    /// <summary>
    /// The prompt (system message) for performing the steps.
    /// </summary>
    private readonly string _stepPrompt;

    /// <summary>
    /// The name to use when creating semantic functions that are restricted from plan creation
    /// </summary>
    private const string StepwisePlannerPluginName = "StepwisePlanner_Excluded";

    /// <summary>
    /// The user message to add to the chat history for each step of the plan.
    /// </summary>
    private const string StepwiseUserMessage = "Perform the next step of the plan if there is more work to do. When you have reached a final answer, use the UserInteraction-SendFinalAnswer function to communicate this back to the user.";

    // Context variable keys
    private const string AvailableFunctionsKey = "available_functions";
    private const string InitialPlanKey = "initial_plan";
    private const string GoalKey = "goal";
    private const string NameDelimiterKey = "name_delimiter";

    #endregion private

    /// <summary>
    /// Plugin used by the <see cref="FunctionCallingStepwisePlanner"/> to interact with the caller.
    /// </summary>
    public sealed class UserInteraction
    {
        /// <summary>
        /// This function is used by the <see cref="FunctionCallingStepwisePlanner"/> to indicate when the final answer has been found.
        /// </summary>
        /// <param name="answer">The final answer for the plan.</param>
        [KernelFunction]
        [Description("This function is used to send the final answer of a plan to the user.")]
#pragma warning disable IDE0060 // Remove unused parameter. The parameter is purely an indication to the LLM and is not intended to be used.
        public string SendFinalAnswer([Description("The final answer")] string answer)
#pragma warning restore IDE0060
        {
            return "Thanks";
        }
    }
}

File: python/semantic_kernel/planners/function_calling_stepwise_planner/function_calling_stepwise_planner.py
# Copyright (c) Microsoft. All rights reserved.

import asyncio
import logging
import os
from copy import copy
from typing import Any

import yaml

from semantic_kernel.connectors.ai.function_calling_utils import kernel_function_metadata_to_function_call_format
from semantic_kernel.connectors.ai.function_choice_behavior import FunctionChoiceBehavior
from semantic_kernel.connectors.ai.open_ai.prompt_execution_settings.open_ai_prompt_execution_settings import (
    OpenAIChatPromptExecutionSettings,
)
from semantic_kernel.connectors.ai.open_ai.services.azure_chat_completion import AzureChatCompletion
from semantic_kernel.connectors.ai.open_ai.services.open_ai_chat_completion import OpenAIChatCompletion
from semantic_kernel.const import DEFAULT_SERVICE_NAME
from semantic_kernel.contents.chat_history import ChatHistory
from semantic_kernel.contents.function_call_content import FunctionCallContent
from semantic_kernel.contents.function_result_content import FunctionResultContent
from semantic_kernel.contents.text_content import TextContent
from semantic_kernel.exceptions.planner_exceptions import PlannerInvalidConfigurationError
from semantic_kernel.functions.kernel_arguments import KernelArguments
from semantic_kernel.functions.kernel_function import KernelFunction
from semantic_kernel.kernel import Kernel
from semantic_kernel.kernel_pydantic import KernelBaseModel
from semantic_kernel.planners.function_calling_stepwise_planner.function_calling_stepwise_planner_options import (
    FunctionCallingStepwisePlannerOptions,
)
from semantic_kernel.planners.function_calling_stepwise_planner.function_calling_stepwise_planner_result import (
    FunctionCallingStepwisePlannerResult,
    UserInteraction,
)
from semantic_kernel.prompt_template.kernel_prompt_template import KernelPromptTemplate
from semantic_kernel.prompt_template.prompt_template_config import PromptTemplateConfig

CUR_DIR = os.path.dirname(os.path.realpath(__file__))
PLAN_YAML_FILE_PATH = os.path.join(CUR_DIR, "generate_plan.yaml")
STEP_PROMPT_FILE_PATH = os.path.join(CUR_DIR, "step_prompt.txt")

STEPWISE_PLANNER_PLUGIN_NAME = "StepwisePlanner_Excluded"

STEPWISE_USER_MESSAGE = (
    "Perform the next step of the plan if there is more work to do. "
    "When you have reached a final answer, use the UserInteraction-SendFinalAnswer "
    "function to communicate this back to the user."
)

USER_INTERACTION_SEND_FINAL_ANSWER = "UserInteraction-SendFinalAnswer"

logger: logging.Logger = logging.getLogger(__name__)


class FunctionCallingStepwisePlanner(KernelBaseModel):
    """A Function Calling Stepwise Planner."""

    service_id: str
    options: FunctionCallingStepwisePlannerOptions
    generate_plan_yaml: str
    step_prompt: str

    def __init__(self, service_id: str, options: FunctionCallingStepwisePlannerOptions | None = None):
        """Initialize a new instance of the FunctionCallingStepwisePlanner.

        The FunctionCallingStepwisePlanner is a planner based on top of an OpenAI Chat Completion service
        (whether it be AzureOpenAI or OpenAI), so that we can use tools.

        If the options are configured to use callbacks to get the initial plan and the step prompt,
        the planner will use those provided callbacks to get that information. Otherwise, it will
        read from the default yaml plan file and the step prompt file.

        Args:
            service_id (str): The service id
            options (Optional[FunctionCallingStepwisePlannerOptions], optional): The options for
                the function calling stepwise planner. Defaults to None.
        """
        options = options or FunctionCallingStepwisePlannerOptions()

        if options.get_initial_plan:
            generate_plan_yaml = options.get_initial_plan()
        else:
            with open(PLAN_YAML_FILE_PATH) as f:
                generate_plan_yaml = f.read()

        if options.get_step_prompt:
            step_prompt = options.get_step_prompt()
        else:
            with open(STEP_PROMPT_FILE_PATH) as f:
                step_prompt = f.read()

        options.excluded_plugins.add(STEPWISE_PLANNER_PLUGIN_NAME)

        super().__init__(
            service_id=service_id,
            options=options,
            generate_plan_yaml=generate_plan_yaml,
            step_prompt=step_prompt,
        )

    async def invoke(
        self,
        kernel: Kernel,
        question: str,
        arguments: KernelArguments | None = None,
        **kwargs: Any,
    ) -> FunctionCallingStepwisePlannerResult:
        """Execute the function calling stepwise planner.

        Args:
            kernel: The kernel instance
            question: The input question
            arguments: (optional) The kernel arguments
            kwargs: (optional) Additional keyword arguments

        Returns:
            FunctionCallingStepwisePlannerResult: The result of the function calling stepwise planner

        Raises:
            PlannerInvalidConfigurationError: If the input question is empty
        """
        if not question:
            raise PlannerInvalidConfigurationError("Input question cannot be empty")

        if not arguments:
            arguments = KernelArguments(**kwargs)

        try:
            chat_completion: OpenAIChatCompletion | AzureChatCompletion = kernel.get_service(service_id=self.service_id)
        except Exception as exc:
            raise PlannerInvalidConfigurationError(
                f"The OpenAI service `{self.service_id}` is not available. Please configure the AI service."
            ) from exc

        if not isinstance(chat_completion, (OpenAIChatCompletion, AzureChatCompletion)):
            raise PlannerInvalidConfigurationError(
                f"The service with id `{self.service_id}` is not an OpenAI based service."
            )

        prompt_execution_settings: OpenAIChatPromptExecutionSettings = (
            self.options.execution_settings
            or chat_completion.instantiate_prompt_execution_settings(service_id=self.service_id)
        )
        if self.options.max_completion_tokens:
            prompt_execution_settings.max_tokens = self.options.max_completion_tokens

        # Clone the kernel so that we can add planner-specific plugins without affecting the original kernel instance
        cloned_kernel = copy(kernel)
        cloned_kernel.add_plugin(UserInteraction(), "UserInteraction")

        # Create and invoke a kernel function to generate the initial plan
        initial_plan = await self._generate_plan(question=question, kernel=cloned_kernel, arguments=arguments)

        chat_history_for_steps = await self._build_chat_history_for_step(
            goal=question, initial_plan=initial_plan, kernel=cloned_kernel, arguments=arguments, service=chat_completion
        )
        prompt_execution_settings.function_choice_behavior = FunctionChoiceBehavior.Auto(
            auto_invoke=False, filters={"excluded_plugins": list(self.options.excluded_plugins)}
        )
        for i in range(self.options.max_iterations):
            # sleep for a bit to avoid rate limiting
            if i > 0:
                await asyncio.sleep(self.options.min_iteration_time_ms / 1000.0)  # convert ms to sec
            # For each step, request another completion to select a function for that step
            chat_history_for_steps.add_user_message(STEPWISE_USER_MESSAGE)
            chat_result = await chat_completion.get_chat_message_contents(
                chat_history=chat_history_for_steps,
                settings=prompt_execution_settings,
                kernel=cloned_kernel,
            )
            chat_result = chat_result[0]
            chat_history_for_steps.add_message(chat_result)

            if not any(isinstance(item, FunctionCallContent) for item in chat_result.items):
                chat_history_for_steps.add_user_message("That function call is invalid. Try something else!")
                continue

            # Try to get the final answer out
            function_call_content = next(
                (
                    item
                    for item in chat_result.items
                    if isinstance(item, FunctionCallContent) and item.name == USER_INTERACTION_SEND_FINAL_ANSWER
                ),
                None,
            )

            if function_call_content is not None:
                args = function_call_content.parse_arguments()
                return FunctionCallingStepwisePlannerResult(
                    final_answer=args.get("answer", ""),
                    chat_history=chat_history_for_steps,
                    iterations=i + 1,
                )

            for item in chat_result.items:
                if not isinstance(item, FunctionCallContent):
                    continue
                try:
                    context = await chat_completion._process_function_call(
                        function_call=item,
                        kernel=cloned_kernel,
                        chat_history=chat_history_for_steps,
                        arguments=arguments,
                        function_call_count=1,
                        request_index=0,
                        function_call_behavior=prompt_execution_settings.function_choice_behavior,
                    )
                    if context is not None:
                        # Only add the function result content to the chat history if the context is present
                        # which means it wasn't added in the _process_function_call method
                        frc = FunctionResultContent.from_function_call_content_and_result(
                            function_call_content=item, result=context.function_result
                        )
                        chat_history_for_steps.add_message(message=frc.to_chat_message_content())
                except Exception as exc:
                    frc = FunctionResultContent.from_function_call_content_and_result(
                        function_call_content=item,
                        result=TextContent(text=f"An error occurred during planner invocation: {exc}"),
                    )
                    chat_history_for_steps.add_message(message=frc.to_chat_message_content())
                    continue

        # We're done, but the model hasn't returned a final answer.
        return FunctionCallingStepwisePlannerResult(
            final_answer="",
            chat_history=chat_history_for_steps,
            iterations=i + 1,
        )

    async def _build_chat_history_for_step(
        self,
        goal: str,
        initial_plan: str,
        kernel: Kernel,
        arguments: KernelArguments,
        service: OpenAIChatCompletion | AzureChatCompletion,
    ) -> ChatHistory:
        """Build the chat history for the stepwise planner."""
        ChatHistory()
        additional_arguments = KernelArguments(
            goal=goal,
            initial_plan=initial_plan,
        )
        arguments.update(additional_arguments)
        kernel_prompt_template = KernelPromptTemplate(
            prompt_template_config=PromptTemplateConfig(
                template=self.step_prompt,
            )
        )
        prompt = await kernel_prompt_template.render(kernel, arguments)
        return ChatHistory.from_rendered_prompt(prompt)

    def _create_config_from_yaml(self, kernel: Kernel) -> "KernelFunction":
        """A temporary method to create a function from the yaml file.

        The yaml.safe_load will be replaced with the proper kernel
        method later.
        """
        data = yaml.safe_load(self.generate_plan_yaml)
        prompt_template_config = PromptTemplateConfig(**data)
        if DEFAULT_SERVICE_NAME in prompt_template_config.execution_settings:
            settings = prompt_template_config.execution_settings.pop(DEFAULT_SERVICE_NAME)
            prompt_template_config.execution_settings[self.service_id] = settings
        return kernel.add_function(
            function_name="create_plan",
            plugin_name="sequential_planner",
            description="Create a plan for the given goal",
            prompt_template_config=prompt_template_config,
        )

    async def _generate_plan(
        self,
        question: str,
        kernel: Kernel,
        arguments: KernelArguments,
    ) -> str:
        """Generate the plan for the given question using the kernel."""
        generate_plan_function = self._create_config_from_yaml(kernel)
        functions_manual = [
            kernel_function_metadata_to_function_call_format(f)
            for f in kernel.get_list_of_function_metadata({
                "excluded_functions": [f"{self.service_id}", "sequential_planner-create_plan"]
            })
        ]
        generated_plan_args = KernelArguments(
            name_delimiter="-",
            available_functions=functions_manual,
            goal=question,
        )
        generated_plan_args.update(arguments)
        generate_plan_result = await kernel.invoke(generate_plan_function, generated_plan_args)
        return str(generate_plan_result)
